### ISSUE TITLE
als crashed when images come in too fast

### ISSUE THREAD

**Issue by fuherb** [gehelem/als#158]
Hi, I try the nightly build 2021-12-10-7533ab0 on a Raspberry Pi 3B+.  Knowing that the Rpi 3B+ has limited computing power, I tried to see how fast it can handle the live stack.  I dumped the jpg image files one by one into the scan folder manually, the first 3 with an interval of about 30s, then I increased to about 20s.  The jpg files are of size about 1M (2028*1540).  At first it stacked OK, then when I shorted the intervals and came to about the 11 image, it crashed.  I copied the terminal output and looked at the steps to check what went wrong.  I attached the txt file for your reference.  I added a [#] to the end of each line to identify which image was being processed, starting from 0 till 10.  I found that when I reduced the dumping interval to about 20s, the process steps started to overlap (ie processing 2 images simultaneously).  It still worked until image #10.  It seems that the program is still saving the post processed stack result of #9 and #10 was kind of process to a "mature" state.  When the saving of #9 was completed, the program "forgot" that it had processed #10 to a certain state already and re-read the file again for processing, then it crashed.  I think the problem is that the program confuses itself by dealing with the same image twice.  Hope this helps.
[als-analysis.txt](https://github.com/gehelem/als/files/7869477/als-analysis.txt)

**Comment by deufrai** [comment]
Hi

Sorry for the late reply ! And many thanks for the effort you did put to provide all relevent info at hand.
This issue is bugging us for a long time and we thought we fixed it for good.

Almost sure it is a memory-related issue.

Thanks for raising the issue

**Comment by fuherb** [comment]
I am a bit surprised but delighted to finally hear the feedback, I thought this project was dead.

Nevertheless it seems that development work will resume?  Really looking forward to this.

When you said this is a memory issue, do you mean that the Raspberry Pi does not have sufficient memory to handle this, or the way ALS handle memory?  I am now using ASTAP in Rpi 3B+ to do live stacking.  It is working fine but ASTAP is not really designed for live stacking.  The way that ASTAP handle rapid in-rush of images is by adding an extra mark to the end of the file extension of the images already processed.  Eg. a jpg file that was stacked will be labelled as .jpg~.  In this way the program will not restack them again.  After the stack is finished, a checkbox is provided to convert all label extensions back to original.  I think this is a good solution to this problem.

Also, is it possible to accept more different file types for live stacking.  Native Raspberry Pi cameras capture RAW images in DNG but ALS does not accept this.  Converting DNG to FIT will take up resources in Pi, which is already limited.  Using JPG seriously degrades the quality.

Great to see ALS moving forward again.

**Comment by deufrai** [comment]
Hi

Main issue, crashing when new images arrive faster then we are able to process them, is the result of a too candid design choice. This will be handled somehow

Regarding the issue of double processing of the same image : as it should not cause any real damage, it should not happen anyway and may be a consequence of the main problem

You could help me get to the bottom of this by providing detailed logs. They can be activated in the preferences and require a restart to take effect.

If you could reproduce the crash and provide the generated logs, we could then dive deeper into this

**Comment by deufrai** [comment]
As for DNG support, I just created #167. Thanks

**Comment by fuherb** [comment]
Thanks for quick reply and the initiative to deal with DNG inputs.

It will be rather difficult for me to re-generate the crash now.  I recently did an update of the Pi OS to the latest Bullseye and that triggered a complete rework of all my image capturing scripts as Bullseye completely abandon the previous camera control commands.  There is no backward compatibility.  As a result, ALS is not installed in my current system.  I do hestitate to make further addition or changes now.  Furthermore I do only capture in DNG now and that's another hurdle to cross.  I think the log that I provided in the start of this thread is about the best I can do now.  When things stablized a bit in my new system that I certainly will install ALS again.  For EAA I think ALS is currently the most suitable app to use in the Pi platform.
